{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8608ee81-b84d-489d-8589-fc25c0ba93f3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-08T08:13:51.871960109Z",
     "start_time": "2023-06-08T08:13:49.226045617Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_47603/1540613272.py:5: UserWarning: Shapely 2.0 is installed, but because PyGEOS is also installed, GeoPandas will still use PyGEOS by default for now. To force to use and test Shapely 2.0, you have to set the environment variable USE_PYGEOS=0. You can do this before starting the Python process, or in your code before importing geopandas:\n",
      "\n",
      "import os\n",
      "os.environ['USE_PYGEOS'] = '0'\n",
      "import geopandas\n",
      "\n",
      "In a future release, GeoPandas will switch to using Shapely by default. If you are using PyGEOS directly (calling PyGEOS functions on geometries from GeoPandas), this will then stop working and you are encouraged to migrate from PyGEOS to Shapely 2.0 (https://shapely.readthedocs.io/en/latest/migration_pygeos.html).\n",
      "  import geopandas as gpd\n"
     ]
    }
   ],
   "source": [
    "import branca\n",
    "import dask.distributed\n",
    "import folium\n",
    "import folium.plugins\n",
    "import geopandas as gpd\n",
    "import numpy as np\n",
    "import rasterio as rio\n",
    "import rasterio.features\n",
    "import shapely.geometry\n",
    "import xrspatial.multispectral as ms\n",
    "from branca.element import Element, Figure\n",
    "from IPython.display import HTML, display\n",
    "from odc.stac import configure_rio, stac_load\n",
    "from pystac_client import Client\n",
    "\n",
    "from utils import convert_bounds, image_on_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "461a6023-257a-4787-b378-cf88e3d0e709",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-08T08:13:54.523393899Z",
     "start_time": "2023-06-08T08:13:54.516113956Z"
    }
   },
   "outputs": [],
   "source": [
    "# stack configuration\n",
    "cfg = {\n",
    "    \"sentinel-s2-l2a-cogs\": {\n",
    "        \"assets\": {\n",
    "            \"*\": {\"data_type\": \"uint16\", \"nodata\": 0},\n",
    "            \"SCL\": {\"data_type\": \"uint8\", \"nodata\": 0},\n",
    "            \"visual\": {\"data_type\": \"uint8\", \"nodata\": 0},\n",
    "        },\n",
    "        \"aliases\": {\"red\": \"B04\", \"green\": \"B03\", \"blue\": \"B02\"},\n",
    "    },\n",
    "    \"*\": {\"warnings\": \"ignore\"},\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f9485798-07fd-4984-b2c7-eb741478f850",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2023-06-08T08:59:31.392352804Z",
     "start_time": "2023-06-08T08:59:30.151840559Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/node.py:182: UserWarning: Port 8787 is already in use.\n",
      "Perhaps you already have a cluster running?\n",
      "Hosting the HTTP server on port 39131 instead\n",
      "  warnings.warn(\n",
      "2023-06-08 10:59:30,491 - distributed.nanny - ERROR - Failed to initialize worker\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,503 - distributed.nanny - ERROR - Failed to initialize worker\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,511 - distributed.nanny - ERROR - Failed to start process\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,516 - distributed.nanny - ERROR - Failed to start process\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,516 - distributed.nanny - ERROR - Failed to initialize worker\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,519 - distributed.nanny - ERROR - Failed to initialize worker\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,519 - distributed.nanny - ERROR - Failed to initialize worker\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,522 - distributed.nanny - ERROR - Failed to initialize worker\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,563 - distributed.nanny - ERROR - Failed to start process\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,564 - distributed.nanny - ERROR - Failed to start process\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,565 - distributed.nanny - ERROR - Failed to start process\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "2023-06-08 10:59:30,566 - distributed.nanny - ERROR - Failed to start process\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-1235099' coro=<_wrap_awaitable() done, defined at /home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py:118> exception=RuntimeError('Nanny failed to start.')>\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 536, in start\n",
      "    @property\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/utils.py\", line 1849, in wait_for\n",
      "    def filter(self, record: logging.LogRecord) -> bool:\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/asyncio/tasks.py\", line 408, in wait_for\n",
      "    return await fut\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 368, in start_unsafe\n",
      "    return self\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py\", line 119, in _wrap_awaitable\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 544, in start\n",
      "    def status(self, value: Status) -> None:\n",
      "RuntimeError: Nanny failed to start.\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-1235097' coro=<_wrap_awaitable() done, defined at /home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py:118> exception=RuntimeError('Nanny failed to start.')>\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 536, in start\n",
      "    @property\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/utils.py\", line 1849, in wait_for\n",
      "    def filter(self, record: logging.LogRecord) -> bool:\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/asyncio/tasks.py\", line 408, in wait_for\n",
      "    return await fut\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 368, in start_unsafe\n",
      "    return self\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py\", line 119, in _wrap_awaitable\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 544, in start\n",
      "    def status(self, value: Status) -> None:\n",
      "RuntimeError: Nanny failed to start.\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-1235095' coro=<_wrap_awaitable() done, defined at /home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py:118> exception=RuntimeError('Nanny failed to start.')>\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 536, in start\n",
      "    @property\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/utils.py\", line 1849, in wait_for\n",
      "    def filter(self, record: logging.LogRecord) -> bool:\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/asyncio/tasks.py\", line 408, in wait_for\n",
      "    return await fut\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 368, in start_unsafe\n",
      "    return self\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py\", line 119, in _wrap_awaitable\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 544, in start\n",
      "    def status(self, value: Status) -> None:\n",
      "RuntimeError: Nanny failed to start.\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-1235100' coro=<_wrap_awaitable() done, defined at /home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py:118> exception=RuntimeError('Nanny failed to start.')>\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 536, in start\n",
      "    @property\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/utils.py\", line 1849, in wait_for\n",
      "    def filter(self, record: logging.LogRecord) -> bool:\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/asyncio/tasks.py\", line 408, in wait_for\n",
      "    return await fut\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 368, in start_unsafe\n",
      "    return self\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py\", line 119, in _wrap_awaitable\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 544, in start\n",
      "    def status(self, value: Status) -> None:\n",
      "RuntimeError: Nanny failed to start.\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-1235098' coro=<_wrap_awaitable() done, defined at /home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py:118> exception=RuntimeError('Nanny failed to start.')>\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 536, in start\n",
      "    @property\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/utils.py\", line 1849, in wait_for\n",
      "    def filter(self, record: logging.LogRecord) -> bool:\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/asyncio/tasks.py\", line 408, in wait_for\n",
      "    return await fut\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 368, in start_unsafe\n",
      "    return self\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py\", line 119, in _wrap_awaitable\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 544, in start\n",
      "    def status(self, value: Status) -> None:\n",
      "RuntimeError: Nanny failed to start.\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-1235096' coro=<_wrap_awaitable() done, defined at /home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py:118> exception=RuntimeError('Nanny failed to start.')>\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 536, in start\n",
      "    @property\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/utils.py\", line 1849, in wait_for\n",
      "    def filter(self, record: logging.LogRecord) -> bool:\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/asyncio/tasks.py\", line 408, in wait_for\n",
      "    return await fut\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 368, in start_unsafe\n",
      "    return self\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py\", line 119, in _wrap_awaitable\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 544, in start\n",
      "    def status(self, value: Status) -> None:\n",
      "RuntimeError: Nanny failed to start.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unexpected exception formatting exception. Falling back to standard exception\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3505, in run_code\n",
      "    if async_:\n",
      "  File \"/tmp/ipykernel_47603/1817239109.py\", line 2, in <module>\n",
      "    client = dask.distributed.Client()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/client.py\", line 1008, in __init__\n",
      "    }\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/client.py\", line 1205, in start\n",
      "    self._loop_runner.start()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/utils.py\", line 416, in sync\n",
      "    if error:\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/utils.py\", line 389, in f\n",
      "    future = wait_for(future, callback_timeout)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/tornado/gen.py\", line 769, in run\n",
      "    # Save the exception for later. It's important that\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/client.py\", line 1271, in _start\n",
      "    address = cfg[\"address\"]\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py\", line 413, in _\n",
      "    def __await__(self: Self) -> Generator[Any, Any, Self]:\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/deploy/spec.py\", line 382, in _correct_state_internal\n",
      "    await asyncio.wait(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 525, in start\n",
      "    raise e\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/core.py\", line 536, in start\n",
      "    @property\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/utils.py\", line 1849, in wait_for\n",
      "    def filter(self, record: logging.LogRecord) -> bool:\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/asyncio/tasks.py\", line 408, in wait_for\n",
      "    return await fut\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 368, in start_unsafe\n",
      "    return self\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 442, in instantiate\n",
      "    plugin = pickle.loads(plugin)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 718, in start\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 866, in _wait_until_connected\n",
      "    msg = self.init_result_q.get_nowait()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/nanny.py\", line 926, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker.py\", line 756, in __init__\n",
      "    self.memory_manager = WorkerMemoryManager(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/worker_memory.py\", line 166, in __init__\n",
      "    self.data = SpillBuffer(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 95, in __init__\n",
      "    slow = Slow(spill_directory, max_spill)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/spill.py\", line 283, in __init__\n",
      "    compression = get_compression_settings(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/distributed/protocol/compression.py\", line 110, in get_compression_settings\n",
      "    name = dask.config.get(key)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/dask/config.py\", line 563, in get\n",
      "    result = result[k]\n",
      "KeyError: 'spill-compression'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 2102, in showtraceback\n",
      "    if hasattr(value, \"_render_traceback_\"):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 1310, in structured_traceback\n",
      "    Minimal='')\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 1199, in structured_traceback\n",
      "    else:\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 1052, in structured_traceback\n",
      "    orig_etype = etype\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 978, in format_exception_as_a_whole\n",
      "    width = min(75, get_terminal_size()[0])\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 878, in format_record\n",
      "    )\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 712, in lines\n",
      "    def show_exception_only(self, etype, evalue):\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/stack_data/utils.py\", line 144, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/stack_data/core.py\", line 734, in lines\n",
      "    pieces = self.included_pieces\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/stack_data/utils.py\", line 144, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/stack_data/core.py\", line 681, in included_pieces\n",
      "    pos = scope_pieces.index(self.executing_piece)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/stack_data/utils.py\", line 144, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/stack_data/core.py\", line 660, in executing_piece\n",
      "    return only(\n",
      "  File \"/home/biel/miniconda3/envs/geo/lib/python3.10/site-packages/executing/executing.py\", line 190, in only\n",
      "    raise NotOneValueFound('Expected one value, found 0')\n",
      "executing.executing.NotOneValueFound: Expected one value, found 0\n"
     ]
    }
   ],
   "source": [
    "# start the dask cluster\n",
    "client = dask.distributed.Client()\n",
    "configure_rio(cloud_defaults=True, aws={\"aws_unsigned\": True}, client=client)\n",
    "client"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e1937ed-40b9-4721-828a-d843e56f449b",
   "metadata": {},
   "source": [
    "# Select a bounding box for your query\n",
    "Use the rectangle tool and then click on it to get the recatngle json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ecbe1d36-bca2-4b3d-8205-6844edce6f9e",
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2023-06-08T08:14:02.413459029Z",
     "start_time": "2023-06-08T08:14:02.399148788Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "<branca.element.Figure at 0x7f2be0689810>",
      "text/html": "<iframe srcdoc=\"&lt;!DOCTYPE html&gt;\n&lt;html&gt;\n&lt;head&gt;\n    \n    &lt;meta http-equiv=&quot;content-type&quot; content=&quot;text/html; charset=UTF-8&quot; /&gt;\n    \n        &lt;script&gt;\n            L_NO_TOUCH = false;\n            L_DISABLE_3D = false;\n        &lt;/script&gt;\n    \n    &lt;style&gt;html, body {width: 100%;height: 100%;margin: 0;padding: 0;}&lt;/style&gt;\n    &lt;style&gt;#map {position:absolute;top:0;bottom:0;right:0;left:0;}&lt;/style&gt;\n    &lt;script src=&quot;https://cdn.jsdelivr.net/npm/leaflet@1.9.3/dist/leaflet.js&quot;&gt;&lt;/script&gt;\n    &lt;script src=&quot;https://code.jquery.com/jquery-1.12.4.min.js&quot;&gt;&lt;/script&gt;\n    &lt;script src=&quot;https://cdn.jsdelivr.net/npm/bootstrap@5.2.2/dist/js/bootstrap.bundle.min.js&quot;&gt;&lt;/script&gt;\n    &lt;script src=&quot;https://cdnjs.cloudflare.com/ajax/libs/Leaflet.awesome-markers/2.0.2/leaflet.awesome-markers.js&quot;&gt;&lt;/script&gt;\n    &lt;link rel=&quot;stylesheet&quot; href=&quot;https://cdn.jsdelivr.net/npm/leaflet@1.9.3/dist/leaflet.css&quot;/&gt;\n    &lt;link rel=&quot;stylesheet&quot; href=&quot;https://cdn.jsdelivr.net/npm/bootstrap@5.2.2/dist/css/bootstrap.min.css&quot;/&gt;\n    &lt;link rel=&quot;stylesheet&quot; href=&quot;https://netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css&quot;/&gt;\n    &lt;link rel=&quot;stylesheet&quot; href=&quot;https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.2.0/css/all.min.css&quot;/&gt;\n    &lt;link rel=&quot;stylesheet&quot; href=&quot;https://cdnjs.cloudflare.com/ajax/libs/Leaflet.awesome-markers/2.0.2/leaflet.awesome-markers.css&quot;/&gt;\n    &lt;link rel=&quot;stylesheet&quot; href=&quot;https://cdn.jsdelivr.net/gh/python-visualization/folium/folium/templates/leaflet.awesome.rotate.min.css&quot;/&gt;\n    \n            &lt;meta name=&quot;viewport&quot; content=&quot;width=device-width,\n                initial-scale=1.0, maximum-scale=1.0, user-scalable=no&quot; /&gt;\n            &lt;style&gt;\n                #map_4e2324ae954c68c4d55b6b2fd523e5e4 {\n                    position: relative;\n                    width: 100.0%;\n                    height: 100.0%;\n                    left: 0.0%;\n                    top: 0.0%;\n                }\n                .leaflet-container { font-size: 1rem; }\n            &lt;/style&gt;\n        \n    &lt;script src=&quot;https://cdnjs.cloudflare.com/ajax/libs/leaflet.draw/1.0.2/leaflet.draw.js&quot;&gt;&lt;/script&gt;\n    &lt;link rel=&quot;stylesheet&quot; href=&quot;https://cdnjs.cloudflare.com/ajax/libs/leaflet.draw/1.0.2/leaflet.draw.css&quot;/&gt;\n&lt;/head&gt;\n&lt;body&gt;\n    \n    \n            &lt;div class=&quot;folium-map&quot; id=&quot;map_4e2324ae954c68c4d55b6b2fd523e5e4&quot; &gt;&lt;/div&gt;\n        \n&lt;/body&gt;\n&lt;script&gt;\n    \n    \n            var map_4e2324ae954c68c4d55b6b2fd523e5e4 = L.map(\n                &quot;map_4e2324ae954c68c4d55b6b2fd523e5e4&quot;,\n                {\n                    center: [0, 0],\n                    crs: L.CRS.EPSG3857,\n                    zoom: 1,\n                    zoomControl: true,\n                    preferCanvas: false,\n                }\n            );\n\n            \n\n        \n    \n            var tile_layer_746f70a60e994a081d9578d52df9bca5 = L.tileLayer(\n                &quot;https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png&quot;,\n                {&quot;attribution&quot;: &quot;Data by \\u0026copy; \\u003ca target=\\&quot;_blank\\&quot; href=\\&quot;http://openstreetmap.org\\&quot;\\u003eOpenStreetMap\\u003c/a\\u003e, under \\u003ca target=\\&quot;_blank\\&quot; href=\\&quot;http://www.openstreetmap.org/copyright\\&quot;\\u003eODbL\\u003c/a\\u003e.&quot;, &quot;detectRetina&quot;: false, &quot;maxNativeZoom&quot;: 18, &quot;maxZoom&quot;: 18, &quot;minZoom&quot;: 0, &quot;noWrap&quot;: false, &quot;opacity&quot;: 1, &quot;subdomains&quot;: &quot;abc&quot;, &quot;tms&quot;: false}\n            ).addTo(map_4e2324ae954c68c4d55b6b2fd523e5e4);\n        \n    \n            var options = {\n              position: &quot;topleft&quot;,\n              draw: {&quot;polyline&quot;: {&quot;allowIntersection&quot;: false}},\n              edit: {&quot;poly&quot;: {&quot;allowIntersection&quot;: false}},\n            }\n            // FeatureGroup is to store editable layers.\n            var drawnItems = new L.featureGroup().addTo(\n                map_4e2324ae954c68c4d55b6b2fd523e5e4\n            );\n            options.edit.featureGroup = drawnItems;\n            var draw_control_1d80d4a9facf78ccb786caaea26f3047 = new L.Control.Draw(\n                options\n            ).addTo( map_4e2324ae954c68c4d55b6b2fd523e5e4 );\n            map_4e2324ae954c68c4d55b6b2fd523e5e4.on(L.Draw.Event.CREATED, function(e) {\n                var layer = e.layer,\n                    type = e.layerType;\n                var coords = JSON.stringify(layer.toGeoJSON());\n                layer.on(&#x27;click&#x27;, function() {\n                    alert(coords);\n                    console.log(coords);\n                });\n                drawnItems.addLayer(layer);\n             });\n            map_4e2324ae954c68c4d55b6b2fd523e5e4.on(&#x27;draw:created&#x27;, function(e) {\n                drawnItems.addLayer(e.layer);\n            });\n            \n        \n&lt;/script&gt;\n&lt;/html&gt;\" width=\"600px\" height=\"600px\"style=\"border:none !important;\" \"allowfullscreen\" \"webkitallowfullscreen\" \"mozallowfullscreen\"></iframe>"
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fig = branca.element.Figure(width=\"600px\", height=\"600px\")\n",
    "m = folium.Map()\n",
    "fig.add_child(m)\n",
    "draw = folium.plugins.Draw(\n",
    "    position=\"topleft\",\n",
    "    draw_options={\"polyline\": {\"allowIntersection\": False}},\n",
    "    edit_options={\"poly\": {\"allowIntersection\": False}},\n",
    ").add_to(m)\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23a12847-dde9-43d9-9719-2ba5674626ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# paste here the bbox json\n",
    "aoi = {\n",
    "    \"type\": \"Polygon\",\n",
    "    \"coordinates\": [\n",
    "        [\n",
    "            [-61.259766, -64.54844],\n",
    "            [-61.259766, -61.648162],\n",
    "            [-53.085938, -61.648162],\n",
    "            [-53.085938, -64.54844],\n",
    "            [-61.259766, -64.54844],\n",
    "        ]\n",
    "    ],\n",
    "}\n",
    "bbox = rasterio.features.bounds(aoi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "943f59b8-740c-44cf-8a33-4daf2b34121c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-08T08:59:18.376991235Z",
     "start_time": "2023-06-08T08:59:17.832832033Z"
    }
   },
   "outputs": [
    {
     "ename": "Exception",
     "evalue": "invalid datetime component: 2023-4",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mException\u001B[0m                                 Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[28], line 3\u001B[0m\n\u001B[1;32m      1\u001B[0m catalog \u001B[38;5;241m=\u001B[39m Client\u001B[38;5;241m.\u001B[39mopen(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mhttps://earth-search.aws.element84.com/v0\u001B[39m\u001B[38;5;124m\"\u001B[39m)\n\u001B[0;32m----> 3\u001B[0m query \u001B[38;5;241m=\u001B[39m \u001B[43mcatalog\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43msearch\u001B[49m\u001B[43m(\u001B[49m\n\u001B[1;32m      4\u001B[0m \u001B[43m    \u001B[49m\u001B[43mcollections\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43msentinel-s2-l2a-cogs\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m      5\u001B[0m \u001B[43m    \u001B[49m\u001B[43mdatetime\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43m2023-4\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m  \u001B[49m\u001B[38;5;66;43;03m# from / to\u001B[39;49;00m\n\u001B[1;32m      6\u001B[0m \u001B[43m    \u001B[49m\u001B[43mlimit\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;241;43m100\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[1;32m      7\u001B[0m \u001B[43m    \u001B[49m\u001B[43mbbox\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mbbox\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m      8\u001B[0m \u001B[43m    \u001B[49m\u001B[43mquery\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m{\u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43meo:cloud_cover\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[43m{\u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mlt\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m0.1\u001B[39;49m\u001B[43m}\u001B[49m\u001B[43m}\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m      9\u001B[0m \u001B[43m)\u001B[49m\n\u001B[1;32m     11\u001B[0m items \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mlist\u001B[39m(query\u001B[38;5;241m.\u001B[39mget_items())\n\u001B[1;32m     12\u001B[0m \u001B[38;5;28mprint\u001B[39m(\u001B[38;5;124mf\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mFound: \u001B[39m\u001B[38;5;132;01m{\u001B[39;00m\u001B[38;5;28mlen\u001B[39m(items)\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m datasets\u001B[39m\u001B[38;5;124m\"\u001B[39m)\n",
      "File \u001B[0;32m~/miniconda3/envs/geo/lib/python3.10/site-packages/pystac_client/client.py:459\u001B[0m, in \u001B[0;36mClient.search\u001B[0;34m(self, method, max_items, limit, ids, collections, bbox, intersects, datetime, query, filter, filter_lang, sortby, fields)\u001B[0m\n\u001B[1;32m    454\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[1;32m    455\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mNotImplementedError\u001B[39;00m(\n\u001B[1;32m    456\u001B[0m         \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mNo link with rel=search could be found in this catalog\u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[1;32m    457\u001B[0m     )\n\u001B[0;32m--> 459\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mItemSearch\u001B[49m\u001B[43m(\u001B[49m\n\u001B[1;32m    460\u001B[0m \u001B[43m    \u001B[49m\u001B[43murl\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43msearch_href\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    461\u001B[0m \u001B[43m    \u001B[49m\u001B[43mmethod\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mmethod\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    462\u001B[0m \u001B[43m    \u001B[49m\u001B[43mmax_items\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mmax_items\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    463\u001B[0m \u001B[43m    \u001B[49m\u001B[43mstac_io\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_stac_io\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    464\u001B[0m \u001B[43m    \u001B[49m\u001B[43mclient\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[1;32m    465\u001B[0m \u001B[43m    \u001B[49m\u001B[43mlimit\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mlimit\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    466\u001B[0m \u001B[43m    \u001B[49m\u001B[43mids\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mids\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    467\u001B[0m \u001B[43m    \u001B[49m\u001B[43mcollections\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mcollections\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    468\u001B[0m \u001B[43m    \u001B[49m\u001B[43mbbox\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mbbox\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    469\u001B[0m \u001B[43m    \u001B[49m\u001B[43mintersects\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mintersects\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    470\u001B[0m \u001B[43m    \u001B[49m\u001B[43mdatetime\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mdatetime\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    471\u001B[0m \u001B[43m    \u001B[49m\u001B[43mquery\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mquery\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    472\u001B[0m \u001B[43m    \u001B[49m\u001B[38;5;28;43mfilter\u001B[39;49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43mfilter\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[1;32m    473\u001B[0m \u001B[43m    \u001B[49m\u001B[43mfilter_lang\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mfilter_lang\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    474\u001B[0m \u001B[43m    \u001B[49m\u001B[43msortby\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43msortby\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    475\u001B[0m \u001B[43m    \u001B[49m\u001B[43mfields\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mfields\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    476\u001B[0m \u001B[43m    \u001B[49m\u001B[43mmodifier\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mmodifier\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    477\u001B[0m \u001B[43m\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m~/miniconda3/envs/geo/lib/python3.10/site-packages/pystac_client/item_search.py:296\u001B[0m, in \u001B[0;36mItemSearch.__init__\u001B[0;34m(self, url, method, max_items, stac_io, client, limit, ids, collections, bbox, intersects, datetime, query, filter, filter_lang, sortby, fields, modifier)\u001B[0m\n\u001B[1;32m    290\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mmethod \u001B[38;5;241m=\u001B[39m method\n\u001B[1;32m    291\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mmodifier \u001B[38;5;241m=\u001B[39m modifier\n\u001B[1;32m    293\u001B[0m params \u001B[38;5;241m=\u001B[39m {\n\u001B[1;32m    294\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mlimit\u001B[39m\u001B[38;5;124m\"\u001B[39m: limit,\n\u001B[1;32m    295\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mbbox\u001B[39m\u001B[38;5;124m\"\u001B[39m: \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_format_bbox(bbox),\n\u001B[0;32m--> 296\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mdatetime\u001B[39m\u001B[38;5;124m\"\u001B[39m: \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_format_datetime\u001B[49m\u001B[43m(\u001B[49m\u001B[43mdatetime\u001B[49m\u001B[43m)\u001B[49m,\n\u001B[1;32m    297\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mids\u001B[39m\u001B[38;5;124m\"\u001B[39m: \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_format_ids(ids),\n\u001B[1;32m    298\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mcollections\u001B[39m\u001B[38;5;124m\"\u001B[39m: \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_format_collections(collections),\n\u001B[1;32m    299\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mintersects\u001B[39m\u001B[38;5;124m\"\u001B[39m: \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_format_intersects(intersects),\n\u001B[1;32m    300\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mquery\u001B[39m\u001B[38;5;124m\"\u001B[39m: \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_format_query(query),\n\u001B[1;32m    301\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mfilter\u001B[39m\u001B[38;5;124m\"\u001B[39m: \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_format_filter(\u001B[38;5;28mfilter\u001B[39m),\n\u001B[1;32m    302\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mfilter-lang\u001B[39m\u001B[38;5;124m\"\u001B[39m: \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_format_filter_lang(\u001B[38;5;28mfilter\u001B[39m, filter_lang),\n\u001B[1;32m    303\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124msortby\u001B[39m\u001B[38;5;124m\"\u001B[39m: \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_format_sortby(sortby),\n\u001B[1;32m    304\u001B[0m     \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mfields\u001B[39m\u001B[38;5;124m\"\u001B[39m: \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_format_fields(fields),\n\u001B[1;32m    305\u001B[0m }\n\u001B[1;32m    307\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_parameters: Dict[\u001B[38;5;28mstr\u001B[39m, Any] \u001B[38;5;241m=\u001B[39m {\n\u001B[1;32m    308\u001B[0m     k: v \u001B[38;5;28;01mfor\u001B[39;00m k, v \u001B[38;5;129;01min\u001B[39;00m params\u001B[38;5;241m.\u001B[39mitems() \u001B[38;5;28;01mif\u001B[39;00m v \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m\n\u001B[1;32m    309\u001B[0m }\n",
      "File \u001B[0;32m~/miniconda3/envs/geo/lib/python3.10/site-packages/pystac_client/item_search.py:511\u001B[0m, in \u001B[0;36mItemSearch._format_datetime\u001B[0;34m(self, value)\u001B[0m\n\u001B[1;32m    509\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m components[\u001B[38;5;241m0\u001B[39m] \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[1;32m    510\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mException\u001B[39;00m(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mcannot create a datetime query with None\u001B[39m\u001B[38;5;124m\"\u001B[39m)\n\u001B[0;32m--> 511\u001B[0m start, end \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_to_isoformat_range\u001B[49m\u001B[43m(\u001B[49m\u001B[43mcomponents\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;241;43m0\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    512\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m end \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[1;32m    513\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;124mf\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mstart\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m/\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mend\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m\"\u001B[39m\n",
      "File \u001B[0;32m~/miniconda3/envs/geo/lib/python3.10/site-packages/pystac_client/item_search.py:464\u001B[0m, in \u001B[0;36mItemSearch._to_isoformat_range\u001B[0;34m(self, component)\u001B[0m\n\u001B[1;32m    462\u001B[0m match \u001B[38;5;241m=\u001B[39m DATETIME_REGEX\u001B[38;5;241m.\u001B[39mmatch(component)\n\u001B[1;32m    463\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m match:\n\u001B[0;32m--> 464\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mException\u001B[39;00m(\u001B[38;5;124mf\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124minvalid datetime component: \u001B[39m\u001B[38;5;132;01m{\u001B[39;00mcomponent\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m\"\u001B[39m)\n\u001B[1;32m    465\u001B[0m \u001B[38;5;28;01melif\u001B[39;00m match\u001B[38;5;241m.\u001B[39mgroup(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mremainder\u001B[39m\u001B[38;5;124m\"\u001B[39m):\n\u001B[1;32m    466\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m match\u001B[38;5;241m.\u001B[39mgroup(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mtz_info\u001B[39m\u001B[38;5;124m\"\u001B[39m):\n",
      "\u001B[0;31mException\u001B[0m: invalid datetime component: 2023-4"
     ]
    }
   ],
   "source": [
    "catalog = Client.open(\"https://earth-search.aws.element84.com/v0\")\n",
    "\n",
    "query = catalog.search(\n",
    "    collections=[\"sentinel-s2-l2a-cogs\"],\n",
    "    datetime=\"2023-4\",  # from / to\n",
    "    limit=100,\n",
    "    bbox=bbox,\n",
    "    query={\"eo:cloud_cover\": {\"lt\": 0.1}},\n",
    ")\n",
    "\n",
    "items = list(query.get_items())\n",
    "print(f\"Found: {len(items)} datasets\")\n",
    "\n",
    "# Convert STAC items into a GeoJSON FeatureCollection\n",
    "stac_json = query.get_all_items_as_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "626bf21a-3be2-4169-bfc8-31cb06834dc0",
   "metadata": {},
   "source": [
    "Show on a map the foodprints of the found datasets "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b4dba49-d0c3-47ad-b4e3-53cc619c9426",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# https://github.com/python-visualization/folium/issues/1501\n",
    "gdf = gpd.GeoDataFrame.from_features(stac_json, \"epsg:4326\")\n",
    "\n",
    "# Compute granule id from components\n",
    "gdf[\"granule\"] = (\n",
    "    gdf[\"sentinel:utm_zone\"].apply(lambda x: f\"{x:02d}\")\n",
    "    + gdf[\"sentinel:latitude_band\"]\n",
    "    + gdf[\"sentinel:grid_square\"]\n",
    ")\n",
    "\n",
    "fig = Figure(width=\"600px\", height=\"600px\")\n",
    "map1 = folium.Map()\n",
    "fig.add_child(map1)\n",
    "\n",
    "folium.GeoJson(\n",
    "    shapely.geometry.box(*bbox),\n",
    "    style_function=lambda x: dict(fill=False, weight=1, opacity=0.7, color=\"olive\"),\n",
    "    name=\"Query\",\n",
    ").add_to(map1)\n",
    "\n",
    "gdf.explore(\n",
    "    \"granule\",\n",
    "    categorical=True,\n",
    "    tooltip=[\n",
    "        \"granule\",\n",
    "        \"datetime\",\n",
    "        \"sentinel:data_coverage\",\n",
    "        \"eo:cloud_cover\",\n",
    "    ],\n",
    "    popup=True,\n",
    "    style_kwds=dict(fillOpacity=0.1, width=2),\n",
    "    name=\"STAC\",\n",
    "    m=map1,\n",
    ")\n",
    "\n",
    "map1.fit_bounds(bounds=convert_bounds(gdf.unary_union.bounds))\n",
    "folium.LayerControl().add_to(map1)\n",
    "\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "327de989-a95b-4732-bd47-446f814451d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "it = [i for i in items if \"20EPQ\" in i.id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "702f966e-6841-4588-a064-4676a0335525",
   "metadata": {},
   "outputs": [],
   "source": [
    "it[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "188cfb71-bcde-4ddb-8939-c2b0e343646b",
   "metadata": {},
   "outputs": [],
   "source": [
    "scene = stac_load(\n",
    "    [it[1]],\n",
    "    bands=[\"B04\", \"B03\", \"B02\"],\n",
    "    crs=\"epsg:3857\",\n",
    "    resolution=50,\n",
    "    chunks={}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d517184-3d79-4163-a359-4ec043453626",
   "metadata": {},
   "outputs": [],
   "source": [
    "scene = scene.where(lambda x: x > 0, other=np.nan).to_array(\"band\").squeeze(\"time\").chunk(\"auto\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a618ce1-8139-4253-95b9-428298ae15ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "scene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bedb501-cd70-4f53-8413-e8d675c98d1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "image_on_map(ms.true_color(*scene).compute(), bbox)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bba2872-de86-4ae2-94e8-0ea9ecc0fd36",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Since we will plot it on a map we need to use `EPSG:3857` projection\n",
    "crs = \"epsg:3857\"\n",
    "\n",
    "data = stac_load(\n",
    "    items,\n",
    "    bands=[\"B04\", \"B03\", \"B02\"],\n",
    "    crs=crs,\n",
    "    resolution=10,\n",
    "    chunks={},  # <-- use Dask\n",
    "    # stac_cfg=cfg,\n",
    "    # bbox=bbox,\n",
    ")\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4e8970a-380a-4d02-a347-787e67a2f2e0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data = (\n",
    "    data.where(lambda x: x > 0, other=np.nan).to_array(\"band\").chunk(\"auto\")\n",
    ")  # sentinel-2 uses 0 as nodata\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19a88b45-6388-4c9c-86d1-426260a26a23",
   "metadata": {},
   "source": [
    "Inspect one of the scenes on the map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0f3a9b5-8c4b-47d6-9878-e05a16dcaea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use ms.true_color to convert RGB chanels and improve the color\n",
    "selection = data.isel(time=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "128f6382-36c4-45f6-ad9f-ea94ee2dc8a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a944fd8-2362-452a-9810-9b1ab86778a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "image_on_map(ms.true_color(*selection).compute(), bbox)\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75e8840c-3a27-4ed6-a125-e81ef4de3ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ms.true_color(*selection).plot.imshow()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f79b458-78c2-4286-830a-58a0cd85c8a4",
   "metadata": {},
   "source": [
    "Compute the cloudless mossaic by taking the median of each pixel across time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49b32d51-667d-4027-ae5b-7254848df261",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "median = data.median(dim=\"time\").compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1398cc0a-5f7d-4a90-a6be-c40039238b71",
   "metadata": {},
   "outputs": [],
   "source": [
    "image = ms.true_color(*median)  # expects red, green, blue DataArrays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5502682a-6de0-49a4-a35d-d4e142811883",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(12, 12))\n",
    "\n",
    "ax.set_axis_off()\n",
    "image.plot.imshow(ax=ax);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39549563-c541-422d-844d-83bb91e75cd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = Figure(width=\"600px\", height=\"600px\")\n",
    "m = folium.Map()\n",
    "fig.add_child(m)\n",
    "image.odc.add_to(m)\n",
    "\n",
    "m.fit_bounds(bounds=convert_bounds(gdf.unary_union.bounds))\n",
    "folium.LayerControl().add_to(m)\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa0de8fd-bfcf-4967-8a1b-9cb5cb6041ee",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
